cocotb/share/lib/embed/gpi_embed.c

```diff
@@ -243,124 +273,163 @@ int embed_sim_init(gpi_sim_info_t *info)
         goto cleanup;
 
     // Obtain the loggpi logger object
-    simlog_obj = PyObject_GetAttrString(cocotb_module, "loggpi");
+    simlog_obj = PyObject_GetAttrString(cocotb_module, "loggpi");       // New reference
 
     if (simlog_obj == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get simlog object from loggpi\n");
+        goto cleanup;
     }
 
-    simlog_func = PyObject_GetAttrString(simlog_obj, "_logFromC");
+    // Obtain the function to use when logging from C code
+    simlog_func = PyObject_GetAttrString(simlog_obj, "_logFromC");      // New reference
     if (simlog_func == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get the _logFromC method");
         goto cleanup;
     }
-
     if (!PyCallable_Check(simlog_func)) {
-        PyErr_Print();
         LOG_ERROR("_logFromC is not callable");
+        Py_DECREF(simlog_func);
         goto cleanup;
     }
 
-    set_log_handler(simlog_func);
-
-    Py_DECREF(simlog_func);
+    set_log_handler(simlog_func);                                       // Note: This function steals a reference to simlog_func.
 
-    simlog_func = PyObject_GetAttrString(simlog_obj, "isEnabledFor");
+    // Obtain the function to check whether to call log function
+    simlog_func = PyObject_GetAttrString(simlog_obj, "isEnabledFor");   // New reference
     if (simlog_func == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get the isEnabledFor method");
         goto cleanup;
     }
-
     if (!PyCallable_Check(simlog_func)) {
-        PyErr_Print();
         LOG_ERROR("isEnabledFor is not callable");
+        Py_DECREF(simlog_func);
         goto cleanup;
     }
 
-    set_log_filter(simlog_func);
+    set_log_filter(simlog_func);                                        // Note: This function steals a reference to simlog_func.
 
-    argv_list = PyList_New(0);
+    // Build argv for cocotb module
+    argv_list = PyList_New(info->argc);                                 // New reference
+    if (argv_list == NULL) {
+        PyErr_Print();
+        LOG_ERROR("Unable to create argv list");
+        goto cleanup;
+    }
     for (i = 0; i < info->argc; i++) {
-        arg_value = PyString_FromString(info->argv[i]);
-        PyList_Append(argv_list, arg_value);
+        PyObject *pArgvItem = PyString_FromString(info->argv[i]);       // New reference
```

Comment:
nit: `arg_item` would be mode consistent with the other locals

Reply:
Changed to `argv_item`.